### PR TITLE
gh-102512: Fix threading after os.fork() called from a foreign thread

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Based on #102517. It recreates a `_DummyThread` as `_MainThread` after fork.

* Issue: gh-102512

### CODE REVIEW

File: Lib/threading.py [hunk 1]
```diff
@@ -1698,6 +1697,9 @@ def _after_fork():
         # by threading.Thread. For example, a thread spawned
         # by thread.start_new_thread().
         current = _MainThread()
+    else:
+        if isinstance(current, _DummyThread):
+            current = _MainThread()
```
Comment (vstinner) [inline Lib/threading.py]: Please add a comment explaining that creating _MainThread() instance overrides `_active[get_ident()]`.

File: Lib/test/test_threading.py [hunk 2]
```diff
@@ -724,6 +724,55 @@ def func():
...
+            from test import support
+
+            def func(lock):
+                # call current_thread() before fork to allocate DummyThread
+                current = threading.current_thread()
+                print(current.name)
+                # flush before fork, so child won't flush it again
+                sys.stdout.flush()
+                pid = os.fork()
+                if pid == 0:
```
Comment (vstinner) [inline Lib/test/test_threading.py:759]: You can start with: `print(type(threading.current_thread()))`? I expect "_MainThread".

File: Lib/test/test_threading.py [hunk 3]
```diff
@@ -724,6 +724,55 @@ def func():
...
+            join_lock = _thread.allocate_lock()
+            join_lock.acquire()
+            th = _thread.start_new_thread(func, (join_lock,))
+            join_lock.acquire()
+        """
+        # "DeprecationWarning: This process is multi-threaded, use of fork() may lead to deadlocks in the child"
+        _, out, err = assert_python_ok("-W", "ignore::DeprecationWarning", "-c", code)
+        data = out.decode().replace('\r', '')
+        self.assertEqual(err, b"")
+        self.assertEqual(data, "Dummy-1\nTrue\nTrue\n")
```
Comment (vstinner) [inline Lib/test/test_threading.py]:
```suggestion
        self.assertEqual(data, 
                         "Dummy-1\n"
                         "True\n"
                         "True\n")
```

**Comment by serhiy-storchaka** [comment]
Thank you for your review Victor. I added more output in tests and added a test for foreign thread without DummyThread.

The code was significantly rewritten. Replacing a DummyThread with a MainThread leaves a DummyThread thread with the same ident (this is disturbing) in `_dangling` set. If there were references to that thread, they still refer to the DummyThread. So it is better to change the type of the existing instance of DummyThread.

File: Lib/test/test_threading.py [hunk 4]
```diff
@@ -722,7 +730,80 @@ def func():
...
+                print(ident in threading._active)
+                # flush before fork, so child won't flush it again
+                sys.stdout.flush()
+                pid = os.fork()
+                if pid == 0:
+                    main = threading.main_thread()
+                    print(main.name, type(main).__name__)
+                    print(main.ident == ident)
+                    print(main is threading.current_thread())
+                    print(main.ident == threading.get_ident())
```
Comment (vstinner) [inline Lib/test/test_threading.py]: You may add prefix for readability:

```suggestion
                    print("name", main.name, type(main).__name__)
                    print("is ident", main.ident == ident)
                    print("is current_thread", main is threading.current_thread())
                    print("get_ident", main.ident == threading.get_ident())
```

File: Lib/test/test_threading.py [hunk 5]
```diff
@@ -722,7 +730,80 @@ def func():
...
+            import os, threading, sys, traceback, _thread
+            from test import support
+
+            def func(lock):
+                ident = threading.get_ident()
+                if %s:
+                    # call current_thread() before fork to allocate DummyThread
+                    current = threading.current_thread()
+                    print(current.name, type(current).__name__)
+                print(ident in threading._active)
```
Comment (vstinner) [inline Lib/test/test_threading.py]: You may add prefix for readability (need to update the expected output):

```suggestion
                print("ident in _active", ident in threading._active)
```

File: Misc/NEWS.d/next/Library/2023-03-08-00-02-30.gh-issue-102512.LiugDr.rst [hunk 6]
```diff
@@ -0,0 +1,4 @@
+When :func:`os.fork` is called from a foreign thread (aka ``_DummyThread``),
+the type of the thread in a child process is changed to ``_MainThread``.
+Also changed its name and daemonic status, it can be now joined.
+
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2023-03-08-00-02-30.gh-issue-102512.LiugDr.rst]: You might remove this empty line. I don't know if it's rendered in HTML?

**Review by vstinner — APPROVED** [review]
LGTM.

I just suggest minor change to make the test easier to read, feel free to ignore.

I dislike setting `self.__class__`, but well, it works as expected and the new tests are exhaustive, so I'm fine with it.

**Pull request merged**